components/gfx/display_list/mod.rs

```diff
@@ -521,26 +336,201 @@ impl DisplayList {
+}
+
+impl<'a> DisplayListTraversal<'a> {
+    pub fn new(display_list: &'a DisplayList) -> DisplayListTraversal {
+        DisplayListTraversal {
+            display_list: display_list,
+            next_item_index: 0,
+            first_item_index: 0,
+            last_item_index: display_list.list.len(),
+        }
+    }
+
+    pub fn new_partial(display_list: &'a DisplayList,
+                       stacking_context_id: StackingContextId,
+                       start: usize,
+                       end: usize)
+                       -> DisplayListTraversal {
+        assert!(display_list.list.len() > start);
+        assert!(display_list.list.len() > end);
+
+        let mut stacking_context_start = start;
+        for x in (0..start).rev()  {
+            if let DisplayItem::PushStackingContextClass(ref item) = display_list.list[x] {
+                if item.stacking_context.id == stacking_context_id {
+                    stacking_context_start = x;
+                    break;
+                }
+            }
+        }
+
+        assert!(stacking_context_start <= start);
+
+        DisplayListTraversal {
+            display_list: display_list,
+            next_item_index: stacking_context_start,
+            first_item_index: start,
+            last_item_index: end + 1,
+        }
+    }
+
+    pub fn previous_item_id(&self) -> usize {
+        return self.next_item_index - 1;
+    }
+
+    pub fn skip_to_end_of_stacking_context(&mut self, id: StackingContextId) {
+        loop {
+            assert!(self.next_item_index < self.last_item_index);
+
+            let item = &self.display_list.list[self.next_item_index];
+            if let &DisplayItem::PopStackingContextClass(ref item) = item {
+                if item.stacking_context_id == id {
+                    return;
+                }
+            }
+            self.next_item_index += 1;
+        }
+    }
+}
+
+impl<'a> Iterator for DisplayListTraversal<'a> {
+    type Item = &'a DisplayItem;
+
+    fn next(&mut self) -> Option<&'a DisplayItem> {
+        loop {
+            if self.next_item_index >= self.last_item_index {
+                return None;
+            }
+
+            assert!(self.next_item_index <= self.last_item_index);
+
+            let reached_first_item = self.next_item_index >= self.first_item_index;
+            let item = &self.display_list.list[self.next_item_index];
+
+            self.next_item_index += 1;
+
+            if reached_first_item {
+                return Some(item)
+            }
+
+            // Before we reach the starting item, we only emit stacking context boundaries. This
+            // is to ensure that we properly position items when we are processing a display list
+            // slice that is relative to a certain stacking context.
+            match item {
+                &DisplayItem::PushStackingContextClass(_) => return Some(item),
+                &DisplayItem::PopStackingContextClass(_) => return Some(item),
```

Comment:
Could use `&DisplayItem::PushStackingContextClass(_) | &DisplayItem::PopStackingContextClass(_) => return Some(item)` to avoid repeating `return`